passqr.1
.TH PASSQR 1 "2014-02-03 +0900" 1.3.1 "Password Store QR add-on"

.SH NAME
passqr - Generate 2D barcodes from Password Store entries


.SH SYNOPSIS
\fBpassqr\fP [\fI OPTIONS ]\fP... \fIPASS-NAME\fB


.SH DESCRIPTION

\fBpassqr\fP is fancy front-end to sending output from \fBpass\fP into
\fBqrencode\fP and piping the output into an image viewer. It facilitates this
process by managing temporary files and providing an interface similar to that
of \fBpass\fP.

The command

.B $ passqr \fIPASS-NAME\fP

will encode the output of \fBpass show PASS-NAME\fP
as a QR code image, then display the image for a short duration on screen.


.SH OPTIONS

.TP
\fB-h, --help\fP
Show usage help and exit.

.TP
\fB-l, --lines\fP \fInum_lines\fP
Encode only the first \fInum_lines\fP lines of output from \fBpass\fP. If
\fInum_lines\fP is 'all', encode all output. If this option is not present, act
as if it was present with argument 1.

.TP
\fB-m, --multiline\fP
Shorthand for \fB--lines all\fP.

.TP
\fB-s, --dotsize\fP
Pass-through for the \fB-s, --size\fP option of \fBqrencode\fP. Default is 6
pixels.

.TP
\fB-t, --timeout\fP \fInum_seconds\fP
Wait \fInum_seconds\fP, which can be a fraction, seconds before closing the
image viewer. Default is 3 seconds.

.TP
\fB--version\fP
Show version information and exit.

.TP
\fB-w, --viewer\fP \fIcommand\fP
Use \fIcommand\fP to display the QR code image. \fIcommand\fP will be appended
with a single filename, and the result will be executed as a command. The
default is taken from the configuration files, see \fbCONFIGURATION\fP.

If you want to print the image to stdout, use \fB-w cat\fP, but note that passqr
makes no guarantee to not print anything else on stdout.


.SH EXAMPLES

.TP
Encode and show in your favourite image viewer
.B alice@laptop ~ $ passqr alice@wonderland.org

.TP
Encode and show in qiv
.B alice@laptop ~ $ passqr -w qiv alice@wonderland.org

.TP
Keep image viewer open for 10 seconds
.B alice@laptop ~ $ passqr -t 10 alice@wonderland.org

.TP
Encode and save image to file
.B alice@laptop ~ $ passqr -w cat alice@wonderland.org > password.png; chmod 400
.B password.png

.TP
Pass environment variables through to \fBpass\fP
.B alice@laptop ~ $ PASSWORD_STORE_DIR=~bob/.password-store passqr
.B bob@cheezburger.com


.SH CONFIGURATION

The configuration files listed under \fBFILES\fP are evaluated as shell scripts
in the order stated above, but only those lines that begin with 'KEY=' where KEY
is one of the configuration keys below.

Later configuration files override earlier ones, and all configuration files are
overridden by commandline arguments.

Of course, this format makes the configuration files attack vectors since they
could easily cause execution of arbitrary code. Don't put stupid stuff in them,
or let anyone else do so.

Your installation probably includes some configuration examples.

.TP
.B Configuration keys

.TP
.B DOTSIZE
Corresponds to the \fB-d, --size\fP option.

.TP
.B TIMEOUT
Corresponds to the \fB-t, --timeout\fP option.

.TP
.B VIEWER_EXEC
Corresponds to the \fB-w, --viewer\fP option.


.SH FILES

.TP
.B /etc/passqr.conf
System-wide configuration. A skeleton version of this is probably included in
your installation.

.TP
.B ${HOME}/.config/passqr.conf
Per-user configuration.


.SH ENVIRONMENT VARIABLES
passqr does not use any environment variables, but all are passed through to \fBpass\fP.

.SH DEPENDENCIES

.IP \(bu 4
Password Store (pass)

.IP \(bu 4
libqrencode

.SH SEE ALSO
.BR pass (1),
.BR qrencode (1)

.SH AUTHOR
\fBpassqr\fP was written by
.MT lundberg.emil@gmail.com
Emil Lundberg
.ME .
The project is developed and hosted publicly on
.UR https://github.com/emlun/passqr
GitHub
.UE . Bug reports, patches and feature requests are welcome in the \fBIssues\fP
section or by e-mail.

.SH COPYING
passqr is free software: you can redistribute it and/or modify it under the
terms of the GNU General Public License as published by the Free Software
Foundation, either version 2 of the License, or (at your option) any later
version.

passqr is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
passqr. If not, see <http://www.gnu.org/licenses/>.


passqr.sh
#!/bin/bash
#
# Copyright 2014 Emil Lundberg <lundberg.emil@gmail.com>. All rights reserved.
#
# This file is part of passqr.
#
# passqr is free software: you can redistribute it and/or modify it under the
# terms of the GNU General Public License as published by the Free Software
# Foundation, either version 2 of the License, or (at your option) any later
# version.
#
# passqr is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along with
# passqr. If not, see <http://www.gnu.org/licenses/>.

GETOPT=getopt
PROGRAM="$(basename "$0")"
VERSION=1.3.1
CONFIG=(
    "/etc/${PROGRAM}.conf"
    "${HOME}/.config/${PROGRAM}.conf"
)

# Default settings, can be overriden by config files and options
DOTSIZE=6
LINES=1
TIMEOUT=3
VERBOSE=false
VIEWER_EXEC=''

# Only evaluate config lines matching this pattern
CONFIG_PATTERN='^(DOTSIZE|TIMEOUT|VIEWER_EXEC)='

for config_file in "${CONFIG[@]}"; do
    if [[ -f "$config_file" ]]; then
        eval $(egrep "$CONFIG_PATTERN" "$config_file")
    fi
done

err() {
    echo "$@" 1>&2
}

usage() {
cat << EOF
Usage: ${PROGRAM} [options] pass-name

Options (defaults):

  -h, --help                    Show this message and exit
  -l, --lines N or 'all' (1)    Encode the first N lines of output from pass
  -m, --multiline               Shorthand for --lines all
  -s, --dotsize N (6)           Dots in generated image are N pixels wide
  -t, --timeout N (3)           Wait N seconds before closing the image viewer
  --version                     Show version information and exit.
  -w, --viewer COMMAND (none)   Use COMMAND to display the QR code image

For details see the man page.

EOF
}

version() {
cat << EOF
Password Store QR add-on, version ${VERSION}
by Emil Lundberg <lundberg.emil@gmail.com>
EOF
}

ARGS="$($GETOPT -o s:hl:mt:w: -l dotsize:,help,lines:,multiline,timeout:,version,viewer: -n "$PROGRAM" -- "$@")"
if [[ $? -ne 0 ]]; then
    usage
    exit 1
fi
eval set -- "$ARGS"

while true; do
    case $1 in
        -h|--help)
            usage
            exit 0
            ;;
        -l|--lines)
            shift
            LINES=$1
            ;;
        -m|--multiline)
            LINES=all
            ;;
        -s|--dotsize)
            shift
            DOTSIZE=$1
            ;;
        -t|--timeout)
            shift
            TIMEOUT=$1
            ;;
        --version)
            version
            exit 0
            ;;
        -w|--viewer)
            shift
            VIEWER_EXEC="$1"
            ;;
        --)
            shift
            break
            ;;
    esac
    shift
done

if [[ $# -eq 0 ]]; then
    # User must pass a pass-name to pass to pass
    err "Fatal: no pass-name given"
    usage
    exit 1
fi

if [[ -z "$VIEWER_EXEC" ]]; then
    err "No image viewer set. Please set VIEWER_EXEC in a config file or use the -w, --viewer 'COMMAND' option. See ${PROGRAM} --help for details."
    err "Recognized config files:"
    for config_file in "${CONFIG[@]}"; do
        err "    $config_file"
    done
    exit 1
fi

output=$(pass show "$@")
passexit=$?
if [[ $passexit -eq 0 ]]; then
    if [[ $LINES != all ]]; then
        output=$(head -n$LINES <<< "$output")
    fi

    tmpfile=$(mktemp --tmpdir "${PROGRAM}.XXXXXXXXXX") || exit $?
    trap "rm '${tmpfile}'" EXIT

    qrencode -s $DOTSIZE -t PNG -o "$tmpfile" "$output" || exit $?
    timeout $TIMEOUT $VIEWER_EXEC "${tmpfile}"

    viewexit=$?
    # Timeout exits with code 124 if command times out
    if [[ $viewexit -eq 124 ]]; then
        exit 0
    else
        exit $viewexit
    fi
else
    err "$output"
    exit $passexit
fi
